### ISSUE TITLE
Investigation: D3D12 does not support SRC_COLOR et al used in SrcBlendAlpha slot

### ISSUE THREAD

**Issue by SenorBlanco** [gpuweb/gpuweb#65]
# D3D12
D3D12 does not support the following enums in the SrcBlendAlpha slot of a D3D12_RENDER_TARGET_BLEND_DESC:

D3D12_BLEND_SRC_COLOR
D3D12_BLEND_INV_SRC_COLOR
D3D12_BLEND_DST_COLOR
D3D12_BLEND_INV_DST_COLOR

"Blend options that end in _COLOR are not allowed." https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/ns-d3d12-d3d12_render_target_blend_desc 

Indeed, at runtime: "D3D12 ERROR: ID3D12Device::CreateBlendState: SrcBlendAlpha[ 0 ] is trying to use a D3D11_BLEND value (0x4) that manipulates color, which is invalid. [ STATE_CREATION ERROR #114: CREATEBLENDSTATE_INVALIDSRCBLENDALPHA]"

# Vulkan
When given *_COLOR in the srcAlphaBlendFactor slot, Vulkan behaves as if the corresponding _ALPHA was passed instead: https://vulkan.lunarg.com/doc/view/1.0.26.0/linux/vkspec.chunked/ch26s01.html#id-1.28.2.17.5

# Metal
Empirically, Metal behaves similarly to Vulkan, but I couldn't find anything in the docs.

# Options
Two possible options: 

1) Have WebGPU also disallow these combinations and error at validation time
2) Make the D3D12 backend do internally what Vulkan/GL/Metal do: map it to corresponding ALPHA variant.

Thoughts?

**Comment by kvark** [comment]
I think the semantics of "as if the corresponding _ALPHA was passed instead" was invented for compatibility with the API model where the user provides the same blend mode for color and alpha (i.e. non-separate). These times are long gone, and we shouldn't be slowed down by that compatibility aspect. Following D3D12 stricter semantics is therefore my preference.

**Comment by kvark** [comment]
I don't have hard preferences here. Following Vulkan and Metal also sounds fine.

**Comment by Kangz** [comment]
Same here, no opinion.

**Comment by Kangz** [comment]
@austinEng hit this today, can we agree to allow `_COLOR` for alpha factors and match Vulkan?

**Comment by kainino0x** [comment]
Weirdly I don't think I ever looked closely at this issue before. I've sort of wondered about this for a long time.

I would personally prefer if the API didn't let you describe nonsense blend modes like `.alpha.srcFactor='src-color'` that have no clear interpretation.

Validation change would be fine but we _could_ even easily describe this at the IDL level by having separate enums for the color and alpha blend modes.

Alternatively, I don't have a good naming suggestion here but if we just replaced the word "color" with something meaning "same component" then the semantics would suddenly make sense, IIUC:

```webidl
enum GPUBlendFactor {
    "zero",
    "one",
    "src-samecomponent",
    "one-minus-src-samecomponent",
    "src-alpha",
    "one-minus-src-alpha",
    "dst-samecomponent",
    "one-minus-dst-samecomponent",
    "dst-alpha",
    "one-minus-dst-alpha",
    "src-alpha-saturated",
    "blend-factor", // tangentially stealing the "blend factor" name from D3D12
    "one-minus-blend-factor"
};

undefined setBlendFactor(GPUColor color);
```

**Comment by kainino0x** [comment]
TIL blendop min/max ignore the blend factor entirely. Probably can't practically make that clear in our API though.

If we decide we want to rename things here, while [looking at](https://www.khronos.org/registry/vulkan/specs/1.2/html/chap27.html#framebuffer-blendfactors) Vulkan just now it occurs to me it could be good to change `src` to `src0` everywhere for explicitness. (Dual source blending adds `src1`.)

And I think we've discussed having holes in the color target array, I wonder how _that_ would interact with blending with src0.

**Comment by Kangz** [comment]
Or just `src-component`? `samecomponent` is pretty heavy.

**Comment by kainino0x** [comment]
Yeah `samecomponent` was just a strawman.

**Comment by kainino0x** [comment]
Was fixed by #1614
